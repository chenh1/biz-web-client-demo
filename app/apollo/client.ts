// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});