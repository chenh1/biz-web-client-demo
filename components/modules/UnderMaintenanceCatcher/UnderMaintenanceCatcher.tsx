'use client'
import React from "react"
import { Section } from "../../containers/Section/Section"
import { Box } from "../../containers/Box/Box"
import { Heading } from "../../core/Heading/Heading"
import { Loader } from "../Loader/Loader"
import { useFlippers } from "../../utils/useFlippers"
import { useSearchParams } from "next/navigation"

interface UnderMaintenanceCatcherProps {
  isOffline?: boolean;
  noData?: boolean;
  error?: string;
  children?: string | JSX.Element | JSX.Element[] | React.ReactNode
}

export const UnderMaintenanceCatcher: React.FC<UnderMaintenanceCatcherProps> = ({
  isOffline = false,
  noData = false,
  error,
  children
}: UnderMaintenanceCatcherProps) => {
  const searchParams = useSearchParams()
  const queryOverride = searchParams?.get('overridemaintenancemode')?.toLowerCase()
  const isOverriding = queryOverride === process.env.QUERY_OVERRIDE
  const { maintenanceMode } = useFlippers({ key: 'maintenanceMode' });
  const isUnavailable = (maintenanceMode && !isOverriding) || error;

  return isUnavailable ?
    isOffline ?
      <Section heightClass="h-screen" className="dark bg-night-black">
        <Box padding="0" className="h-screen relative items-center">
          <Box padding="0" gap="md" className="place-content-center">
            <Heading className="text-center" type="h1">Tending to the flowers right now.</Heading>
            <Heading className="text-center" type="h1">We'll be back shortly!</Heading>
            <Heading className="text-center" type="h4">Thanks for your patience while we're under maintenance. See you soon!</Heading>
            <Loader/>
          </Box>
        </Box>
      </Section>
      :
      noData ?
        <Heading className="text-center" type="h4">We're getting this page up for you right now. Sit tight and thanks for your patience!</Heading>
        :
        <Section heightClass="h-screen" className="dark bg-night-black">
          <Box padding="0" className="h-screen relative items-center">
            <Box padding="0" gap="md" className="place-content-center">
              <Heading className="text-center" type="h1">Tending to the flowers right now.</Heading>
              <Heading className="text-center" type="h1">We'll be back shortly!</Heading>
              <Heading className="text-center" type="h4">Thanks for your patience while we're under maintenance. See you soon!</Heading>
              <Loader/>
            </Box>
          </Box>
        </Section>
      :
      <>
        {children}
      </>
}