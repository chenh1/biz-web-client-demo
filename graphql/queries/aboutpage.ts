import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;