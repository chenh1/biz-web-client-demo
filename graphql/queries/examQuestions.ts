import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`