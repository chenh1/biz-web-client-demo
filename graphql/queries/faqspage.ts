import { gql, DocumentNode } from "@apollo/client";

const faqsPageQuery: DocumentNode = gql`
  query FaqsPage {
    faqsPage {
      data {
        attributes {
          title
          description
          faq(pagination: { page: 1, pageSize: 50 }) {
            question
            answer
            category {
              data {
                attributes {
                  displayName
                  value
                }
              }
            }
            answerRich
          }
        }
      }
    }
  }
`;

export default faqsPageQuery;