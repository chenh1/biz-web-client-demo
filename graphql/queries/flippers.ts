import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
