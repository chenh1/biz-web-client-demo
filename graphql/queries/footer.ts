import { gql, DocumentNode } from "@apollo/client";

export type FooterData = {
  linkGroups: {
    title: string;
    link: {
      url: string;
      copy: string;
    }[]
  }[];
  disclaimer: string;
}

export type FooterDataRes = {
  footer: {
    data: {
      attributes: FooterData
    }
  }
}

export const footerQuery: DocumentNode = gql`
  query Footer {
    footer {
      data {
        attributes {
          linkGroups {
            link {
              url
              copy
            }
            title
          }
          disclaimer
        }
      }
    }
  }
`