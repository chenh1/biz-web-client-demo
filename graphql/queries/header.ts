import { gql, DocumentNode } from "@apollo/client";

export type HeaderData = {
  navLinks: {
    url: string;
    copy: string;
  }[]
}

export type HeaderDataRes = {
  header: {
    data: {
      attributes: HeaderData
    }
  }
}

export const headerQuery: DocumentNode = gql`
  query Header {
    header {
      data {
        attributes {
          navLinks {
            url
            copy
          }
        }
      }
    }
  }
`