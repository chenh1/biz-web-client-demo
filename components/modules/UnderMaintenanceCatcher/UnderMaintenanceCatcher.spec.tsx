import { UnderMaintenanceCatcher } from "./UnderMaintenanceCatcher";
import { MockedProvider } from "@apollo/client/testing";
import { flippersQuery } from "../../../graphql/queries/flippers";
import { fireEvent, render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
// import { resetPasswordMutation } from "../../../graphql/mutations/resetPassword";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


jest.mock("next/navigation", () => ({
  useSearchParams: () => {
    return {
      get: () => "1",
      keys: () => []
    }
  }
}));

const getMocks = (isMaintenanceMode) => ([
  {
    request: {
      query: flippersQuery,
    },
    result: {
      data: {
        flippers: {
          data: {
            attributes: { maintenanceMode: isMaintenanceMode }
          }
        }
      }
    }
  }
])

it('renders a maintenance module in offline state', async () => {
  const { getByTestId, container } = render(
    <MockedProvider mocks={getMocks(true)}>
      <UnderMaintenanceCatcher 
        isOffline={true}
        error={"ReferenceError: "}
      />
    </MockedProvider>
  )

  expect(container).toMatchSnapshot();
});

it('renders a maintenance module in no data state', async () => {
  const { getByTestId, container } = render(
    <MockedProvider mocks={getMocks(true)}>
      <UnderMaintenanceCatcher 
        isOffline={false}
        noData={true}
        error={"ReferenceError: "}
      />
    </MockedProvider>
  )

  expect(container).toMatchSnapshot();
});

it('renders a generic maintenance module', async () => {
  const { getByTestId, container } = render(
    <MockedProvider mocks={getMocks(true)}>
      <UnderMaintenanceCatcher 
        isOffline={false}
        noData={false}
        error={"ReferenceError: "}
      />
    </MockedProvider>
  )

  expect(container).toMatchSnapshot();
});

it('renders the child component since not in maintenance', async () => {
  const { getByTestId, container } = render(
    <MockedProvider mocks={getMocks(false)}>
      <UnderMaintenanceCatcher 
        isOffline={false}
        noData={false}
      >
        <p>Sucessfully loaded</p>
      </UnderMaintenanceCatcher>
    </MockedProvider>
  )

  expect(container).toMatchSnapshot();
});
