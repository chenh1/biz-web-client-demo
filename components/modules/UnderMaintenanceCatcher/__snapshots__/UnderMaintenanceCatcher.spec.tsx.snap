// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a generic maintenance module 1`] = `
<div>
  <div
    class=" flex border-box justify-around w-full h-screen dark bg-night-black"
  >
    <div
      class="w-full max-w-7xl px-4 xl:px-0"
    >
      <div
        class=" grid  gap-4  p-2 h-screen relative items-center"
      >
        <div
          class=" grid  gap-8  p-2 place-content-center"
        >
          <h1
            class="text-night-black dark:text-max-white font-headings font-semibold text-4xl sm:text-6xl text-center"
          >
            Tending to the flowers right now.
          </h1>
          <h1
            class="text-night-black dark:text-max-white font-headings font-semibold text-4xl sm:text-6xl text-center"
          >
            We'll be back shortly!
          </h1>
          <h4
            class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md text-center"
          >
            Thanks for your patience while we're under maintenance. See you soon!
          </h4>
          <div
            class=" grid  gap-4  p-2 slow-spin justify-center"
          >
            <img
              alt="Logo icon"
              height="256"
              src="https://biz-web-client-demo-static.s3.us-east-2.amazonaws.com/biz-web-client-demobloommono_ba249abb01.png"
              width="256"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a maintenance module in no data state 1`] = `
<div>
  <h4
    class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md text-center"
  >
    We're getting this page up for you right now. Sit tight and thanks for your patience!
  </h4>
</div>
`;

exports[`renders a maintenance module in offline state 1`] = `
<div>
  <div
    class=" flex border-box justify-around w-full h-screen dark bg-night-black"
  >
    <div
      class="w-full max-w-7xl px-4 xl:px-0"
    >
      <div
        class=" grid  gap-4  p-2 h-screen relative items-center"
      >
        <div
          class=" grid  gap-8  p-2 place-content-center"
        >
          <h1
            class="text-night-black dark:text-max-white font-headings font-semibold text-4xl sm:text-6xl text-center"
          >
            Tending to the flowers right now.
          </h1>
          <h1
            class="text-night-black dark:text-max-white font-headings font-semibold text-4xl sm:text-6xl text-center"
          >
            We'll be back shortly!
          </h1>
          <h4
            class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md text-center"
          >
            Thanks for your patience while we're under maintenance. See you soon!
          </h4>
          <div
            class=" grid  gap-4  p-2 slow-spin justify-center"
          >
            <img
              alt="Logo icon"
              height="256"
              src="https://biz-web-client-demo-static.s3.us-east-2.amazonaws.com/biz-web-client-demobloommono_ba249abb01.png"
              width="256"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the child component since not in maintenance 1`] = `
<div>
  <p>
    Sucessfully loaded
  </p>
</div>
`;
